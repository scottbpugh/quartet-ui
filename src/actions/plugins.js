// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {createAction} from "redux-actions";

export default {
  addToTreeServers: createAction("PLUGINS_ADD_TO_TREE_SERVERS"),
  removeFromTreeServers: createAction("PLUGINS_REMOVE_FROM_TREE_SERVERS"),
  addButtonToControls: createAction("PLUGINS_ADD_BUTTON_TO_CONTROLS"),
  removeButtonFromControls: createAction("PLUGINS_REMOVE_BUTTON_FROM_CONTROLS"),
  pluginEnabled: createAction("PLUGINS_NEW_PLUGIN_ENABLED"),
  pluginDisabled: createAction("PLUGINS_PLUGIN_DISABLED"),
  receivedPluginsData: createAction("PLUGINS_RECEIVED_PLUGINS_DATA"),
  addLocalPlugin: createAction("PLUGINS_ADD_LOCAL_PLUGIN"),
  pluginListUpdated: createAction("PLUGINS_PLUGIN_LIST_UPDATED"),
  resetPlugins: createAction("PLUGINS_RESET_PLUGINS")
};
window.qu4rtet.exports("actions/plugins", this);
