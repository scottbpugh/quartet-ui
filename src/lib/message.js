// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {Position, Toaster, Intent} from "@blueprintjs/core";
import {pluginRegistry} from "plugins/pluginRegistration";
import stringHash from "string-hash";

const MINIMUM_EXPIRE = 7000;

const msgToaster = Toaster.create({
  className: "my-toaster",
  position: Position.BOTTOM_RIGHT
});

const getIntent = type => {
  switch (type) {
    case "success":
      return Intent.SUCCESS;
    case "danger":
    case "error":
      return Intent.DANGER;
    case "warning":
      return Intent.WARNING;
    default:
      return Intent.PRIMARY;
  }
};

const _msgHashMap = {};

const showMessageNoRepeat = (toastMessage, expires_in = MINIMUM_EXPIRE) => {
  let msgHash = stringHash(JSON.stringify(toastMessage));
  let timestamp = new Date().getTime();
  if (msgHash in _msgHashMap && _msgHashMap[msgHash].expires > timestamp) {
    // don't show msg and update the expiration to prevent the same message
    // over and over.
    _msgHashMap[msgHash].expires = timestamp + expires_in;
    return;
  }
  _msgHashMap[msgHash] = {expires: timestamp + expires_in};
  msgToaster.show(toastMessage);
};

export const showMessage = msg => {
  try {
    // for errors etc out of a component context.
    const intl = pluginRegistry.getIntl();
    if (msg.id) {
      showMessageNoRepeat(
        {
          message: intl.formatMessage({id: msg.id}, msg.values),
          intent: getIntent(msg.type)
        },
        msg.expires_in || MINIMUM_EXPIRE
      );
    } else if (msg.msg) {
      //let formatMsg = typeof msg.msg === "object" ? JSON.stringify(msg.msg) : msg.msg;
      showMessageNoRepeat(
        {message: msg.msg, intent: getIntent(msg.type)},
        msg.expires_in || MINIMUM_EXPIRE
      );
    }
  } catch (e) {
    console.log(
      "an error occurred while attempting to display an error toast.",
      e
    );
  }
};

window.qu4rtet.exports("lib/message", this);
