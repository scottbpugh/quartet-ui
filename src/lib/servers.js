// Copyright (c) 2018 Serial Lab
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import Swagger from "swagger-client";
import {pluginRegistry} from "plugins/pluginRegistration";
import {store} from "store";
import actions from "actions/serversettings";

/**
 * Server - Holds data about a server and its settings, API client, ...
 */
export class Server {
  constructor(serverSettings) {
    /*
    Following object should be passed.
    {
      serverSettingName,
      hostname,
      port,
      ssl,
      serverID,
      path,
      username,
      password
    }
    */
    this.setServerData(serverSettings);
    // make saved server object available to core and plugins.
    pluginRegistry.registerServer(this);
  }

  setServerData = serverSettings => {
    /*
    Following object should be passed.
    {
      serverSettingName,
      hostname,
      port,
      ssl,
      serverID,
      path,
      username,
      password
    */
    // allow for partial updates.
    for (let key in serverSettings) {
      if (serverSettings[key] !== "" || serverSettings[key] !== undefined) {
        this[key] = serverSettings[key];
      }
    }
    this.protocol = this.ssl === true ? "https" : "http";
    this.port = !this.port ? 80 : this.port;
    this.path = !this.path && !serverSettings.path ? "" : this.path;
    this.url = this.getServerURL();
    this._client = null;
    this.manifest = null;
    this.appList = [];
  };
  getFormStructure = () => {
    return Server.getFormStructure(this.toJSON());
  };
  static getFormStructure = (initialValues = {}) => [
    {
      name: "serverSettingName",
      defaultValue: initialValues.serverSettingName || null,
      description: {
        type: "text",
        required: true,
        read_only: false,
        label: "Server Setting Name",
        help_text:
          "The label that will be used for this server connection setting."
      }
    },
    {
      name: "serverID",
      defaultValue: initialValues.serverID || null,
      description: {
        type: "hidden",
        required: false,
        read_only: false,
        label: "Server ID",
        help_text: "Hidden Server ID"
      }
    },
    {
      name: "hostname",
      defaultValue: initialValues.hostname || null,
      description: {
        type: "text",
        required: true,
        read_only: false,
        label: "Server Hostname",
        help_text:
          "A hostname or IP address, example localhost, serial-box.com, or 192.168.5.10."
      }
    },
    {
      name: "port",
      defaultValue: initialValues.port || null,
      description: {
        type: "number",
        required: true,
        read_only: false,
        label: "Port Number",
        min_value: 1,
        max_value: 65535,
        help_text: "A port to connect to. Example, 80, 8080, 443, ..."
      }
    },
    {
      name: "path",
      defaultValue: initialValues.path || "",
      description: {
        type: "text",
        required: false,
        read_only: false,
        label: "Root Path",
        helperText: "A path to interact with API (Optional), example /api"
      }
    },
    {
      name: "ssl",
      defaultValue: initialValues.ssl || false,
      description: {
        type: "boolean",
        required: false,
        read_only: false,
        label: "SSL/TLS",
        helperText: "SSL/TLS encryption"
      }
    },
    {
      name: "username",
      defaultValue: initialValues.username || null,
      description: {
        type: "text",
        required: false,
        read_only: false,
        label: "Username",
        help_text: "Basic Auth Username"
      }
    },
    {
      name: "password",
      defaultValue: initialValues.password || null,
      description: {
        type: "password",
        required: false,
        read_only: false,
        label: "Password",
        help_text: "Basic Auth Password"
      }
    }
  ];
  getArrayFields = () => {
    return [
      {name: "protocol", value: this.protocol, editable: true},
      {name: "port", value: this.port, editable: true},
      {name: "path", value: this.path, editable: true},
      {name: "ssl", value: this.ssl, editable: true},
      {name: "url", value: this.url, editable: false},
      {name: "hostname", value: this.hostname, editable: true},
      {
        name: "serverSettingsName",
        value: this.serverSettingName,
        editable: true
      },
      {name: "username", value: this.username},
      {name: "serverID", value: this.serverID}
    ];
  };

  toJSON() {
    return {
      serverID: this.serverID,
      protocol: this.protocol,
      port: this.port,
      path: this.path,
      ssl: this.ssl,
      hostname: this.hostname,
      serverSettingName: this.serverSettingName,
      url: this.url,
      appList: this.appList,
      username: this.username,
      password: this.password
    };
  }

  getServerURL = () => {
    return `${this.protocol}://${this.hostname}:${this.port}/${this.path}`;
  };

  parseSchema = () => {
    return new Promise((resolve, reject) => {
      Swagger(this.url + "schema/")
        .then(client => {
          // swagger-js client is available.
          // client.spec / client.originalSpec / client.errors
          resolve(client);
        })
        .catch(error => {
          reject(error);
        });
    });
  };

  listApps = () => {
    if (this.appList.length === 0) {
      this.getClient()
        .then(client => {
          this.appList = Object.keys(client.apis);
          // let redux know we got our data
          store.dispatch({
            type: actions.appsListUpdated,
            payload: this.toJSON()
          });
        })
        .catch(error => {
          // handle legacy.
          this.appList = ["serialbox"];
          // let redux know we got our data
          store.dispatch({
            type: actions.appsListUpdated,
            payload: this.toJSON()
          });
        });
    }
  };

  getClient = () => {
    return new Promise((resolve, reject) => {
      if (this._client) {
        resolve(this._client);
      } else {
        // we don't have a client yet.
        // Fetch it.
        this.parseSchema()
          .then(client => {
            this._client = client;
            resolve(this._client);
          })
          .catch(error => {
            reject(error);
          });
      }
    });
  };

  getManifest = () => {
    new Promise((resolve, reject) => {
      if (!this.manifest) {
        this.getClient().then(client => {
          client.apis.manifest.quartet_manifest_list().then(result => {
            this.manifest = result.body;
            resolve(result.body);
          });
        });
      } else {
        resolve(this.manifest);
      }
    });
  };
}
