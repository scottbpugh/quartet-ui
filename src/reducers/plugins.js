// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import {handleActions} from "redux-actions";
import actions from "actions/plugins";
import {showMessage} from "lib/message";

export const initialData = () => {
  return {
    navTreeItems: [],
    plugins: {}
  };
};

export const setEnablePlugin = pluginEntries => {
  return dispatch => {
    for (let plugin in pluginEntries) {
      pluginEntries[plugin].enabled = true;
    }
    showMessage({type: "success", msg: `Plugin enabled.`});
    return dispatch({type: actions.pluginEnabled, payload: pluginEntries});
  };
};
export const setDisablePlugin = pluginEntries => {
  return dispatch => {
    for (let plugin in pluginEntries) {
      pluginEntries[plugin].enabled = false;
    }
    showMessage({type: "success", msg: "Plugin disabled."});
    return dispatch({type: actions.pluginDisabled, payload: pluginEntries});
  };
};
export default handleActions(
  {
    [actions.addToTreeServers]: (state, action) => {
      return {
        ...state,
        navTreeItems: {
          ...state.navTreeItems,
          [action.payload.pluginComponentName]: {...action.payload}
        }
      };
    },
    [actions.removeFromTreeServers]: (state, action) => {
      let navTreeItems = {...state.navTreeItems};
      delete navTreeItems[action.payload.pluginComponentName];
      return {
        ...state,
        navTreeItems: navTreeItems
      };
    },
    [actions.pluginEnabled]: (state, action) => {
      return {
        ...state,
        plugins: {...state.plugins, ...action.payload}
      };
    },
    [actions.pluginDisabled]: (state, action) => {
      return {
        ...state,
        plugins: {...state.plugins, ...action.payload}
      };
    }
  },
  {}
);
