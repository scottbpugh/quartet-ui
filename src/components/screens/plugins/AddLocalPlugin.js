// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React, {Component} from "react";
import {Field, reduxForm} from "redux-form";
import {FormGroup, Intent, Button} from "@blueprintjs/core";
import {RightPanel} from "components/layouts/Panels";
import {FormattedMessage} from "react-intl";
import {connect} from "react-redux";
import {addLocalPlugin} from "reducers/plugins";

const FORM_NAME = "addPlugin";

class _AddLocalPlugin extends Component {
  submit = postValues => {
    // add the flag it's a local plugin.
    postValues["local"] = true;
    postValues["core"] = false;
    this.props.addLocalPlugin(postValues);
    this.props.history.push("/plugins");
  };
  render() {
    return (
      <RightPanel title={<FormattedMessage id="app.plugins.addPluginForm" />}>
        <form onSubmit={this.props.handleSubmit(this.submit.bind(this))}>
          <div>
            <FormGroup
              helperText="The name of the plugin, for programmatic use. Alphanumeric with dashes and underscores."
              label="Plugin Name"
              require={true}>
              <Field
                className="bp3-input"
                name="pluginName"
                component="input"
                type="text"
                placeholder="Plugin Name"
              />
            </FormGroup>
            <FormGroup
              helperText="The path to initialize the plugin. The file containing enable and disable."
              label="Init Path"
              require={true}>
              <Field
                className="bp3-input"
                name="initPath"
                component="input"
                type="text"
                placeholder="Init Path"
              />
            </FormGroup>
            <FormGroup
              helperText="The path to the preview thumbnail"
              label="preview">
              <Field
                className="bp3-input"
                name="preview"
                component="input"
                type="text"
                placeholder="Preview Path"
              />
            </FormGroup>
            <FormGroup
              helperText="The name as it will appear in the plugin list"
              label="Readable Name">
              <Field
                className="bp3-input"
                name="readableName"
                component="input"
                type="text"
                placeholder="Readable Name"
                require={true}
              />
            </FormGroup>
            <FormGroup
              helperText="The package path or package name in NPM."
              label="Package Path">
              <Field
                className="bp3-input"
                name="packagePath"
                component="input"
                type="text"
                placeholder="Package Path or NPM package name"
                require={true}
              />
            </FormGroup>
            <FormGroup
              helperText="The description of what this plugin does."
              label="Description">
              <Field
                className="bp3-input"
                name="description"
                component="input"
                type="text"
                placeholder="Description"
                require={true}
              />
            </FormGroup>
          </div>
          <Button type="submit">Submit</Button>
        </form>
      </RightPanel>
    );
  }
}

export default connect(
  (state, ownProps) => {
    initialValues: ownProps.initialValues || {};
  },
  {addLocalPlugin}
)(reduxForm({form: FORM_NAME})(_AddLocalPlugin));
