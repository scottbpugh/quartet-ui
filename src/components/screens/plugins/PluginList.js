// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React, {Component} from "react";
import {RightPanel} from "components/layouts/Panels";
import {Card, Button, Tag} from "@blueprintjs/core";
import {
  setEnablePlugin,
  setDisablePlugin,
  fetchRemotePlugins
} from "reducers/plugins";
import {connect} from "react-redux";
import "./PluginList.css";
import {updateMessages} from "reducers/locales";

class Plugin extends Component {
  constructor(props) {
    super(props);
    this.state = {runtimeVersion: this.getVersionIfInstalled()};
  }
  componentDidMount() {
    this.setState({runtimeVersion: this.getVersionIfInstalled()});
  }
  componentWillReceiveProps(nextProps) {
    this.setState({runtimeVersion: this.getVersionIfInstalled()});
  }
  handleEnable = async evt => {
    let pluginModule = await window.qu4rtet.getPluginModule(
      this.props.pluginEntry
    );
    pluginModule.enablePlugin();
    this.props.setEnablePlugin({
      [this.props.pluginName]: {...this.props.pluginEntry}
    });
    this.props.updateMessages(this.props.locale);
  };
  getVersionIfInstalled = () => {
    try {
      if (this.props.pluginEntry.packagePath) {
        return qu4rtet.require(
          `${qu4rtet.pluginPath}/${
            this.props.pluginEntry.packagePath
          }/package.json`
        ).version;
      }
    } catch (e) {
      // we don't care if it's not available.
      return null;
    }
  };
  handleDisable = async evt => {
    let pluginModule = await window.qu4rtet.getPluginModule(
      this.props.pluginEntry
    );
    pluginModule.disablePlugin();
    this.props.setDisablePlugin({
      [this.props.pluginName]: {...this.props.pluginEntry}
    });
  };
  render() {
    return (
      <Card className="pt-elevation-4">
        <h5>
          {this.props.pluginEntry.readableName}{" "}
          {this.props.pluginEntry.local ? (
            <Button
              iconName="pt-icon-edit"
              className="pt-button add-plugin-button pt-intent-primary"
              onClick={this.handleEnable.bind(this)}>
              Edit
            </Button>
          ) : null}
          {!this.props.pluginEntry || !this.props.pluginEntry.enabled ? (
            <Button
              iconName="pt-icon-add"
              className="pt-button add-plugin-button pt-intent-primary"
              onClick={this.handleEnable.bind(this)}>
              Enable
            </Button>
          ) : (
            <Button
              iconName="pt-icon-remove"
              className="pt-button add-plugin-button pt-intent-primary"
              onClick={this.handleDisable.bind(this)}>
              Disable
            </Button>
          )}
        </h5>
        <div>
          {this.state.runtimeVersion ? (
            <Tag>{`version ${this.state.runtimeVersion}`}</Tag>
          ) : null}
          <div className="pt-callout pt-intent-primary pt-callout-plugin">
            {this.props.pluginEntry.preview ? (
              <img
                className="plugin-screenshot"
                src={this.props.pluginEntry.preview}
                title={this.props.pluginEntry.readableName}
                alt={this.props.pluginEntry.readableName}
              />
            ) : null}
            <p>{this.props.pluginEntry.description}</p>
          </div>
        </div>
      </Card>
    );
  }
}

export class _PluginList extends Component {
  constructor(props) {
    super(props);
    this.state = {plugins: {}};
  }
  componentDidMount() {
    this.props.fetchRemotePlugins();
  }
  componentWillReceiveProps(nextProps) {
    this.setState({plugins: nextProps.plugins});
  }
  render() {
    return (
      <RightPanel title={<formattedMessage id="app.nav.plugins" />}>
        <div className="cards-container">
          {this.state.plugins
            ? Object.keys(this.state.plugins).map(pluginName => {
                return (
                  <Plugin
                    {...this.props}
                    key={pluginName}
                    pluginName={pluginName}
                    pluginEntry={this.state.plugins[pluginName]}
                  />
                );
              })
            : null}
        </div>
      </RightPanel>
    );
  }
}

export const PluginList = connect(
  (state, ownProps) => {
    return {
      plugins: state.plugins.plugins,
      locale: state.intl.locale
    };
  },
  {setEnablePlugin, setDisablePlugin, updateMessages, fetchRemotePlugins}
)(_PluginList);
