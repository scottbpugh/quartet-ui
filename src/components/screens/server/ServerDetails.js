// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import React, {Component} from "react";
import {RightPanel} from "components/layouts/Panels";
import {FormattedMessage} from "react-intl";
import {CopyToClipboard} from 'react-copy-to-clipboard';
import {connect} from "react-redux";
import {saveServer} from "reducers/serversettings";
import {pluginRegistry} from "plugins/pluginRegistration";
import {Card, Button, Icon, Tag, Intent} from "@blueprintjs/core";
import "./server-details.css";
import {ServerForm} from "./ServerForm";
import {showMessage} from "lib/message";
import classNames from "classnames";
class _ServerDetails extends Component {
  constructor(props) {
    super(props);
    this.state = {
      editMode: false, 
      hidePassword: true,
      copied: false
    };
  }
  componentDidMount() {
    // retrigger a schema fetch when mounting details.
    this.fetchAppList();
  }
  fetchAppList = evt => {
    let serverObject = pluginRegistry.getServer(this.props.server.serverID);
    if (serverObject) {
      serverObject.listApps();
    }
  };
  fetchAppListRefresh = evt => {
    let serverObject = pluginRegistry.getServer(this.props.server.serverID);
    if (serverObject) {
      serverObject.listApps(true);
    }
  };
  toggleEditMode = () => {
    this.setState({editMode: !this.state.editMode});
  };
  submitCallback = () => {
    // leave form on successful submit.
    this.setState({editMode: false});
  };
  unhidePass = (event) => {
    event.preventDefault();
    this.setState({
      hidePassword: !this.state.hidePassword
    })
    console.log(this.state.hidePassword)
  }
  showPopup = (event) => {
    event.preventDefault();
    showMessage({
      id: "Password copied to clipboard",
      type: "success",
    });
  }
  render() {
    let serverObject = pluginRegistry.getServer(this.props.server.serverID);
    let services = serverObject
      ? serverObject.appList
          .filter(service => {
            // remove empty string.
            return service;
          })
          .map(service => {
            return (
              <li key={service}>
                <img  src="./icon.png" />
                {service}
                <span className="icon-dot" />
                {/* <span className="icon-dot" /> */}
              </li>
            );
          })
      : [];
    return (
      <RightPanel title={<FormattedMessage id="app.servers.serverDetails" />}>
        {serverObject ? (
          <div className="cards-container">
            <Card className="pt-elevation-4">
              <h5>
                Settings
                <Button
                  onClick={this.toggleEditMode}
                  className="pt-intent-primary add-incard-button"
                  iconName="pt-icon-edit">
                  Edit
                </Button>
              </h5>
              {this.state.editMode ? (
                <div className="form-card">
                  <ServerForm
                    defaultValues={serverObject.toJSON()}
                    formData={serverObject.getFormStructure()}
                    saveButtonMsg={
                      <FormattedMessage id="app.servers.updateServer" />
                    }
                    submitCallback={this.submitCallback}
                  />
                </div>
              ) : (
                <div>
                  <table className="pt-table pt-bordered pt-striped">
                    <thead>
                      <tr>
                        <th>Server API Endpoint</th>
                        <th>
                          <h6>
                            <Tag intent={Intent.PRIMARY}>
                              <a
                                style={{color: "#FFF"}}
                                href={serverObject.url}
                                target="_blank">
                                {serverObject.url}
                              </a>
                            </Tag>
                          </h6>
                        </th>
                      </tr>
                    </thead>

                    <tbody>
                      {serverObject.getArrayFields().map(elem => {
                        return (
                          <tr key={elem.name} 
                            className={
                              classNames({
                                "hidden-line": elem.hidden === true
                            })}>
                            <td>{elem.name}</td>
                            <td>{"" + elem.value}</td>
                          </tr>
                        );
                      })}
                      {/* <tr>
                        <td>password</td>
                        <td className="flex-row">
                            <div className="password-flex">
                              {this.state.hidePassword === false ? this.props.server.password : "*******"}
                            
                            </div>
                              <div className="button-container">
                              <button
                                tabIndex="0"
                                className={
                                  classNames({
                                    "pt-button pt-icon-eye-open pt-button-inside-array": this.state.hidePassword === true,
                                    "pt-button pt-icon-eye-off button-bg-red red tomato-red pt-button-inside-array": this.state.hidePassword === false
                                  })
                                }
                                onClick={this.unhidePass}
                              >
                              </button> 
                              <CopyToClipboard text={this.props.server.password}
                                onCopy={() => this.setState({copied: true})}>
                                <button 
                                className="pt-button pt-icon-clipboard pt-button-inside-array-2" 
                                onClick={this.showPopup}
                                ></button>
                              </CopyToClipboard>
                            </div>
                        </td>
                      </tr> */}
                    </tbody>
                  </table>
                </div>
              )}
            </Card>
            <Card className="pt-elevation-4">
              <h5>
                Services Enabled{" "}
                <Button
                  onClick={this.fetchAppListRefresh}
                  iconName="pt-icon-refresh add-incard-button"
                />
              </h5>
              {services.length > 0 ? (
                <ul className="service-list">{services}</ul>
              ) : (
                <div onClick={this.fetchAppList} className="centered-action">
                  <Icon
                    iconName="pt-icon-refresh"
                    className="very-large-icon"
                  />
                  <span>Retry</span>
                </div>
              )}
            </Card>
          </div>
        ) : null}
      </RightPanel>
    );
  }
}

export var ServerDetails = connect(
  (state, ownProps) => ({
    server: state.serversettings.servers[ownProps.match.params.serverID]
  }),
  {
    saveServer
  }
)(_ServerDetails);
