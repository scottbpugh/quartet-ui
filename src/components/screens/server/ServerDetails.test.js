// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React, {Component} from "react";
import "tools/mockStore"; // mock ipcRenderer, localStorage, ...
import renderer from "react-test-renderer";
import {ServerDetails} from "./ServerDetails";
import {initialData} from "../../../reducers/serversettings";
import configureStore from "redux-mock-store";
import {MemoryRouter as Router, withRouter} from "react-router-dom";
import {IntlProvider, intlReducer} from "react-intl-redux";
import {Provider} from "react-redux";
import {mockStore, TestWrapper, initialState} from "tools/mockStore";
import {pluginRegistry} from "plugins/pluginRegistration";
import {Server} from "lib/servers";
import sinon from "sinon";
import * as message from "lib/message";

it("renders correctly", () => {
  sinon.stub(message, "showMessage").callsFake(() => {});
  let store = mockStore(initialState);
  const props = {
    match: {
      params: {
        serverID: "d0246781-67c6-474b-8ab0-29de61b6e6bb"
      }
    }
  };
  pluginRegistry.registerServer(
    new Server(
      initialState.serversettings.servers[
        "d0246781-67c6-474b-8ab0-29de61b6e6bb"
      ]
    )
  );
  const serverSettings = renderer
    .create(
      <TestWrapper>
        <Provider store={store}>
          <Router>
            <ServerDetails {...props} store={store} />
          </Router>
        </Provider>
      </TestWrapper>
    )
    .toJSON();
  expect(serverSettings).toMatchSnapshot();
});
