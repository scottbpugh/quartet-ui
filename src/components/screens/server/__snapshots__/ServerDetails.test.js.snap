// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="right-panel"
  >
    <div>
      <div
        className="cards-container"
      >
        <div
          className="bp3-card bp3-elevation-0 bp3-elevation-1"
        >
          <h5 className="bp3-heading">
            Settings
            <button
              className="bp3-button edit bp3-intent-primary add-incard-button"
              onClick={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              type="button"
            >
              <span>
                Edit
              </span>
            </button>
          </h5>
          <div>
            <table
              className="bp3-table bp3-bordered bp3-striped"
            >
              <thead>
                <tr>
                  <th>
                    Server API Endpoint
                  </th>
                  <th>
                    <h6>
                      <span
                        className="bp3-tag bp3-intent-primary"
                      >
                        <a
                          href="http://localhost:8000/"
                          style={
                            Object {
                              "color": "#FFF",
                            }
                          }
                          target="_blank"
                        >
                          http://localhost:8000/
                        </a>
                      </span>
                    </h6>
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td>
                    protocol
                  </td>
                  <td>
                    http
                  </td>
                </tr>
                <tr>
                  <td>
                    port
                  </td>
                  <td>
                    8000
                  </td>
                </tr>
                <tr>
                  <td>
                    path
                  </td>
                  <td>

                  </td>
                </tr>
                <tr>
                  <td>
                    ssl
                  </td>
                  <td>
                    false
                  </td>
                </tr>
                <tr>
                  <td>
                    url
                  </td>
                  <td>
                    http://localhost:8000/
                  </td>
                </tr>
                <tr>
                  <td>
                    hostname
                  </td>
                  <td>
                    localhost
                  </td>
                </tr>
                <tr>
                  <td>
                    serverSettingName
                  </td>
                  <td>
                    box 1
                  </td>
                </tr>
                <tr>
                  <td>
                    authType
                  </td>
                  <td>
                    basic
                  </td>
                </tr>
                <tr>
                  <td>
                    tokenType
                  </td>
                  <td>
                    Token
                  </td>
                </tr>
                <tr>
                  <td>
                    username
                  </td>
                  <td>
                    lduros
                  </td>
                </tr>
                <tr>
                  <td>
                    serverID
                  </td>
                  <td>
                    d0246781-67c6-474b-8ab0-29de61b6e6bb
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div
          className="bp3-card bp3-elevation-0 bp3-elevation-1"
        >
          <h5 className="bp3-heading">
            Services Enabled

            <button
              className="bp3-button refresh add-incard-button"
              onClick={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              type="button"
            />
          </h5>
          <div
            className="centered-action"
            onClick={[Function]}
          >
            <span
              className="standard refresh very-large-icon"
            />
            <span>
              Retry
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
